mitiq/collector.py

```diff
@@ -0,0 +1,263 @@
+
+        try:
+            result = list(result)
+            self._computed_results += result
+            self._executed_circuits += to_run
+        except TypeError:
+            self._computed_results.append(result)
+            self._executed_circuits.append(to_run)
+
+    @staticmethod
+    def is_batched_executor(executor: Callable) -> bool:
+        """Returns True if the input function is recognized as a "batched
+        executor", else False.
+
+        The executor is detected as "batched" if and only if it is annotated
+        with a return type that is one of the following:
+
+            * Iterable[float]
+            * List[float]
+            * Sequence[float]
+            * Tuple[float]
+            * numpy.ndarray
+
+        Batched executors can run several quantum programs in a single call.
+        See below.
+
+        Args:
+            executor: A "serial executor" (1) or a "batched executor" (2).
+
+                (1) A function which inputs a single `QPROGRAM` and outputs a
+                single expectation value as a float.
+                (2) A function which inputs a list of `QPROGRAM`s and outputs a
+                list of expectation values (one for each `QPROGRAM`).
+
+        Returns:
+            True if the executor is detected as batched, else False.
+        """
+        executor_annotation = inspect.getfullargspec(executor).annotations
+
+        return executor_annotation.get("return") in (
+            List[float],
+            Sequence[float],
+            Tuple[float],
+            Iterable[float],
+            np.ndarray,
+        )
+
+
+class CircuitCollection:
+    """A collection of circuits, nominally to execute on some backend."""
+
+    def __init__(self, circuits: Sequence[QPROGRAM]) -> None:
+        """Initializes a CircuitCollection.
+
+        Args:
+            circuits: Sequence of circuits to form the CircuitCollection.
+        """
+        self._raw_circuits = deepcopy(circuits)
+        self._cirq_circuits = [
+            convert_to_mitiq(circuit)[0] for circuit in self._raw_circuits
+        ]
+
+        # List of unique circuits.
+        self._unique: List[cirq.Circuit] = []
+
+        # Dictionary where each key refers to an index of self._raw_circuits
+        # that is the first occurrence of a unique circuit. The value of this
+        # key is the multiplicity/count (number of occurrences) of this circuit
+        # in self._raw_circuits. See CircuitCollection.unique_with_counts.
+        self._counts: Dict[int, int] = {}
+
+        for i, circ in enumerate(self._cirq_circuits):
+            found = False
+            for j, circuit in enumerate(self._unique):
+                if _equal(circ, circuit):
+                    self._counts[list(self._counts.keys())[j]] += 1
+                    found = True
+                    break
+
+            if not found:
+                self._unique.append(circ)
+                self._counts[i] = 1
```

Comment:
This loop could also be simplified if `Circuits` could be hashed. Then a built in `defaultdict` would allow:
The loop that follows could then use a DefaultDict to be:
```
from collections import defaultdict

counts_dict = defaultdict()
for circ in self._cirq_circuits:
    counts_dict[circ] += 1
```

Reply:
This Cirq class may be useful for switching between hashable and unhasbable circuit types:
https://cirq.readthedocs.io/en/latest/generated/cirq.FrozenCircuit.html

Reply:
I think `FrozenCircuit`s with a `Counter` is a nice option.

```python
circuits = [cirq.FrozenCircuit(circ) for circ in all_circuits]
counts = Counter(circuits)
```

For example,

```python
>>> circuits
[0: ───H───,
 0: ───H───,
 0: ───H───,
 0: ───H───,
 0: ───H───,
 0: ───H───,
 0: ───H───,
 0: ───X───,
 0: ───X───,
 0: ───X───,
 0: ───X───,
 0: ───X───]

>>> Counter(circuits)
Counter({0: ───H───: 7, 0: ───X───: 5})
```

Reply:
We'd need to rely on cirq-unstable until cirq v0.10 is released, but this shouldn't be a problem.

Reply:
Nice find! That could simplify our code a lot. Do we still need a CircuitCollection object or can we now drop it and use Counter to make dictionaries?